SpatialGDK/Source/SpatialGDK/Private/Utils/InterestFactory.cpp

```diff
@@ -51,29 +58,90 @@ QueryConstraint InterestFactory::CreateClientCheckoutRadiusConstraint(USpatialCl
-	TArray<QueryConstraint> CheckoutRadiusConstraints = CheckoutRadiusConstraintUtils::BuildNonDefaultActorCheckoutConstraints(
-		DistanceToActorTypeComponents, ClassInfoManager);
+		// The previously built map dedupes spatial constraints. Now the actual query constraints can be built of the form:
+		// OR(AND(cyl(radius), OR(actor 1 components, actor 2 components, ...)), ...)
+		// which is equivalent to having a separate spatial query for each actor type if the radius is the same.
+		TArray<QueryConstraint> CheckoutRadiusConstraints = CheckoutRadiusConstraintUtils::BuildNonDefaultActorCheckoutConstraints(
+			DistanceToActorTypeComponents, ClassInfoManager);
 
-	// Add all the different actor queries to the overall checkout constraint.
-	for (auto& ActorCheckoutConstraint : CheckoutRadiusConstraints)
+		// Add all the different actor queries to the overall checkout constraint.
+		for (auto& ActorCheckoutConstraint : CheckoutRadiusConstraints)
+		{
+			CheckoutRadiusConstraintRoot.OrConstraint.Add(ActorCheckoutConstraint);
+		}
+	}
+	else
 	{
-		CheckoutRadiusConstraint.OrConstraint.Add(ActorCheckoutConstraint);
+		const TArray<float> NetCullDistances = ClassInfoManager->GetNetCullDistances();
+
+		for (float NetCullDistance : NetCullDistances)
+		{
+			const float MaxCheckoutRadiusMeters = CheckoutRadiusConstraintUtils::NetCullDistanceSquaredToSpatialDistance(NetCullDistance);
+			const uint32 NCDComponentId = ClassInfoManager->GetComponentIdForNetCullDistance(NetCullDistance);
+
+			QueryConstraint ComponentConstraint;
+			ComponentConstraint.ComponentConstraint = NCDComponentId;
+
+			if (SpatialGDKSettings->bEnableNetCullDistanceFrequency)
+			{
+				float FullFrequencyCheckoutRadius = MaxCheckoutRadiusMeters * SpatialGDKSettings->FullFrequencyNetCullDistanceRatio;
+
+				{
+					// Add default interest query which doesn't include a frequency
+					QueryConstraint RadiusConstraint;
+					RadiusConstraint.RelativeCylinderConstraint = RelativeCylinderConstraint{ FullFrequencyCheckoutRadius };
+
+					QueryConstraint CheckoutRadiusConstraint;
+					CheckoutRadiusConstraint.AndConstraint.Add(RadiusConstraint);
+					CheckoutRadiusConstraint.AndConstraint.Add(ComponentConstraint);
+
+					CheckoutRadiusConstraintRoot.OrConstraint.Add(CheckoutRadiusConstraint);
+				}
+
+				// Add interest query for specified distance/frequency pairs
+				for (auto DistanceFrequencyPair : SpatialGDKSettings->InterestRangeFrequencyPairs)
+				{
+					float CheckoutRadius = MaxCheckoutRadiusMeters * DistanceFrequencyPair.DistanceRatio;
+
+					QueryConstraint RadiusConstraint;
+					RadiusConstraint.RelativeCylinderConstraint = RelativeCylinderConstraint{ CheckoutRadius };
+
+					QueryConstraint CheckoutRadiusConstraint;
+					CheckoutRadiusConstraint.AndConstraint.Add(RadiusConstraint);
+					CheckoutRadiusConstraint.AndConstraint.Add(ComponentConstraint);
+
+					CheckoutConstraints.Add({ DistanceFrequencyPair.Frequency, CheckoutRadiusConstraint });
```

Comment:
Very much a nit because I would rather the static variable were assigned once if we are using this pattern, but: If we are building this set of `FrequencyConstraint`s here, why not have this function return the separate constraint/frequency pairs and instead build the full query (tack on the pre-computed result type) a layer up and cache the queries?

Reply:
Full query can't be constructed here, as we don't have the rest of the constraints.